### ISSUE TITLE
False positive `reportIncompatibleMethodOverride` with version 1.1.394

### ISSUE THREAD

**Issue by eachimei** [microsoft/pyright#9925]
```python
from collections.abc import Iterator
from contextlib import contextmanager

@contextmanager
def thecontext() -> Iterator[None]:
    yield


class Mixin:
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        print("mixin says hello")


class Concrete(Mixin):

    @thecontext()  # error: "__init__" overrides method of same name in class "Mixin" with incompatible type "_WrappedCallable[(self: Concrete), None]" (reportIncompatibleMethodOverride)
    def __init__(self) -> None:
        super().__init__()
        print("concrete says hello")
```

Versions:
pyright-1.1.394
Python 3.10

\* This issue doesn't happen with pyright 1.1.393

**Comment by erictraut** [comment]
For most methods, it would be correct for pyright to emit a `reportIncompatibleMethodOverride` error in this case. However, the `__init__` and `__new__` methods are exempted from LSP checks, so it should not generate an error in this particular case.

This will be addressed in the next release of pyright.

**Comment by eachimei** [comment]
Thank you! 👍

**Comment by eachimei** [comment]
Can you please elaborate on why it would be correct (to emit the error) on methods other than `__init__` and `__new__`? Asking because `contextmanager` uses [ContextDecorator](https://docs.python.org/3/library/contextlib.html#contextlib.ContextDecorator), which doesn't change the underlying function signature (effectively/from usage perspective). As they say in the docs, it merely acts as a syntactic sugar:

![Image](https://github.com/user-attachments/assets/450acb32-522e-489e-b1a1-df43a368e7b6)

**Comment by erictraut** [comment]
An overridden method must be be assignable to the type of the method it is overriding. Otherwise it violates the LSP (Liskov Substitution Principle). In this case, the decorated type of the override method is not assignable to the type of the method in the base class based on the way `contextmanager` is defined.

```python
from collections.abc import Iterator
from contextlib import contextmanager

@contextmanager
def thecontext() -> Iterator[None]:
    yield

class A:
    def method(self) -> None: ...

class B(A):
    @thecontext()
    def method(self) -> None: ...


reveal_type(A.method)  # (self: A) -> None
reveal_type(B.method)  # _WrappedCallable[(self: B) -> None]

A.method = B.method  # Error: not assignable
```

This situation involves an under-specified behavior within the typing spec regarding the use of a `ParamSpec` when it captures the signature of an instance or class method. (The `contextmanager` call uses a `ParamSpec`.) I won't go into all of the details here, but depending on how this behavior is eventually specified, this may affect the behavior of this sample.

**Comment by eachimei** [comment]
Ah, I see. Thank you, much appreciated!

**Comment by eachimei** [comment]
Found sightings of the issue with a bugfix in typeshed from last month:

https://github.com/python/typeshed/pull/13416

https://discuss.python.org/t/defining-overload-preserving-signatures-for-wrapped-callables/78350

I assume that the above typeshed changeset is not yet reflected in pyright 1.1.394?
With the typeshed fix, is pyright expected to work "better" with ContextDecorator?

**Comment by JacobHayes** [comment]
I noticed a similar issue in [Logfire](https://github.com/pydantic/logfire), which uses a `@contextmanager`/`ContextDecorator` around an `__exit__` method [here](https://github.com/pydantic/logfire/blob/a42341514f0dc19255354519e75bcd0a4b9f281a/logfire/_internal/main.py#L2110-L2113).

The code works fine at runtime and pyright is content with the `__enter__`/`__exit__` definition (which seems aligned with [this comment](https://github.com/microsoft/pyright/issues/9694#issuecomment-2588318686)), but flags about an invalid context manager when used (eg: in user code).

I'm happy to open a new issue if you think this is separate, but it seems like the root cause is shared changes to how `ContextDecorator` is interpreted?

```python
from collections.abc import Generator
from contextlib import contextmanager
from typing import Any

# from logfire._internal.utils import handle_internal_errors


@contextmanager
def handle_internal_errors() -> Generator[None]:
    try:
        print("before handle_internal_errors yield")
        yield
        print("after handle_internal_errors yield")
    except Exception:
        print("in handle_internal_errors except")
    print("after handle_internal_errors try/except")


class Foo:
    def __enter__(self) -> None:
        print("enter")

    @handle_internal_errors()
    def __exit__(
        self, exc_type: type[BaseException] | None, exc_value: BaseException | None, traceback: Any
    ) -> None:
        print("exit")


with Foo():  # pyright: Object of type "Foo" cannot be used with "with" because it does not correctly implement __exit__
    print("inside context")

# enter
# inside context
# inside handle_internal_errors
# before handle_internal_errors yield
# exit
# after handle_internal_errors yield
# after handle_internal_errors try/except
```

```sh
$ pyright logfire_internal_errors.py
WARNING: there is a new pyright version available (v1.1.393 -> v1.1.394).
Please install the new version or set PYRIGHT_PYTHON_FORCE_VERSION to `latest`

0 errors, 0 warnings, 0 informations

$ pip3 install 'pyright==1.1.394'
[...]

$ pyright logfire_internal_errors.py
[...]/logfire_internal_errors.py
  [...]/logfire_internal_errors.py:29:6 - error: Object of type "Foo" cannot be used with "with" because it does not correctly implement __exit__ (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**Comment by benglewis** [comment]
@erictraut This looks to be the same issue as in #9920

**Comment by erictraut** [comment]
@benglewis, no this isn't at all related to #9920.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.395.
